target/linux/ramips/dts/MIR3P.dts

```diff
@@ -0,0 +1,180 @@
+#include "mt7621.dtsi"
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+
+/ {
+	compatible = "xiaomi,mir3p", "mediatek,mt7621-soc";
+	model = "Xiaomi Mi Router 3 Pro";
+
+	aliases {
+		led-boot = &led_status_yellow;
+		led-failsafe = &led_status_red;
+		led-running = &led_status_blue;
+		led-upgrade = &led_status_yellow;
+	};
+
+	memory@0 {
+		device_type = "memory";
+		reg = <0x0 0x1c000000>, <0x20000000 0x04000000>;
+	};
+
+	chosen {
+		bootargs = "console=ttyS0,115200n8";
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		led_status_red: status_red {
+			label = "mir3p:red:status";
+			gpios = <&gpio0 6 GPIO_ACTIVE_LOW>;
+		};
+
+		led_status_blue: status_blue {
+			label = "mir3p:blue:status";
+			gpios = <&gpio0 8 GPIO_ACTIVE_LOW>;
+		};
+
+		led_status_yellow: status_yellow {
+			label = "mir3p:yellow:status";
+			gpios = <&gpio0 10 GPIO_ACTIVE_LOW>;
+		};
+
+		wan_amber {
+			label = "mir3p:amber:wan";
+			gpios = <&gpio0 14 GPIO_ACTIVE_LOW>;
+		};
+
+		lan3_amber {
+			label = "mir3p:amber:lan3";
+			gpios = <&gpio0 13 GPIO_ACTIVE_LOW>;
+		};
+
+		lan2_amber {
+			label = "mir3p:amber:lan2";
+			gpios = <&gpio0 16 GPIO_ACTIVE_LOW>;
+		};
+
+		lan1_amber {
+			label = "mir3p:amber:lan1";
+			gpios = <&gpio0 15 GPIO_ACTIVE_LOW>;
+		};
+	};
+
+	button {
+		compatible = "gpio-keys-polled";
+		poll-interval = <20>;
+
+		reset {
+			label = "reset";
+			gpios = <&gpio0 18 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_RESTART>;
+		};
+	};
+
+	reg_usb_vbus: regulator {
+		compatible = "regulator-fixed";
+		regulator-name = "usb_vbus";
+		regulator-min-microvolt = <5000000>;
+		regulator-max-microvolt = <5000000>;
+		gpio = <&gpio0 12 GPIO_ACTIVE_HIGH>;
+		enable-active-high;
+	};
+};
+
+&xhci {
+	vbus-supply = <&reg_usb_vbus>;
+};
+
+&nand {
+	status = "okay";
+
+	partitions {
+		compatible = "fixed-partitions";
+		#address-cells = <1>;
+		#size-cells = <1>;
+
+		partition@0 {
+			label = "Bootloader";
+			reg = <0x0 0x40000>;
+			read-only;
+		};
+
+		partition@40000 {
+			label = "Config";
+			reg = <0x40000 0x40000>;
+		};
+
+		partition@80000 {
+			label = "Bdata";
+			reg = <0x80000 0x40000>;
+			read-only;
+		};
+
+		factory: partition@c0000 {
+			label = "Factory";
+			reg = <0x0c0000 0x40000>;
+			read-only;
+		};
+
+		partition@100000 {
+			label = "crash";
+			reg = <0x100000 0x40000>;
+		};
+
+		partition@140000 {
+			label = "crash_syslog";
+			reg = <0x140000 0x80000>;
+		};
+
+		partition@1c0000 {
+			label = "reserved0";
+			reg = <0x1c0000 0x40000>;
+			read-only;
+		};
+
+		/* We keep stock xiaomi firmware (kernel0) here */
+		partition@200000 {
+			label = "kernel_stock";
+			reg = <0x200000 0x400000>;
+		};
+
+		firmware {
+			label = "firmware";
+			reg = <0x600000 0xf980000>;
+		};
+
+		partition@600000 {
+			label = "kernel";
+			reg = <0x600000 0x400000>;
+		};
+
+		partition@a00000 {
+			label = "ubi";
+			reg = <0xa00000 0xf580000>;
+		};
+	};
+};
+
+&pcie {
+	/* mt76 doesn't support mt7615 yet... */
```

Comment:
Since you enable the pcie node. Can you please also add the two device nodes, even if both are currently unsupported? 

From what I know the MT7615 does have the pciid 14c3:7615. What I don't know, where the chips are connected
a bootlog could help but I think it should look something like this:

```
&pcie0 {
	wifi@0,0 {
		compatible = "pci14c3,7615";
		reg = <0x0000 0 0 0 0>;
	};
};

&pcie1 {
	wifi@0,0 {
		compatible = "pci14c3,7615";
		reg = <0x0000 0 0 0 0>;
	};
};
```

Reply:
Yeah, good idea. I had originally added them, then saw other DTS files didn't mention them at all, and thought that was for a reason.

Here's what (openwrt) bootlog says:

```
[    1.990415] pci 0000:01:00.0: [14c3:7615] type 00 class 0x000280                                                                    
[    1.990465] pci 0000:01:00.0: reg 0x10: [mem 0x00000000-0x000fffff 64bit]                                                           
[    1.990732] pci_bus 0000:01: busn_res: [bus 01-ff] end is updated to 01                                                             
[    1.990913] pci 0000:02:00.0: [14c3:7615] type 00 class 0x000280                                                                    
[    1.990963] pci 0000:02:00.0: reg 0x10: [mem 0x00000000-0x000fffff 64bit]                                                           
[    1.991236] pci_bus 0000:02: busn_res: [bus 02-ff] end is updated to 02            
```
I'm not sure what you mean when you say "I don't know where the chips are connected"... What I came up with is this (looking at other dts files):
                                                 
```
&pcie0 {
	wifi@0,0 {
		compatible = "pci14c3,7615";
		reg = <0x0000 0 0 0 0>;
	};
};

&pcie1 {
	wifi@0,0 {
		compatible = "pci14c3,7615";
		reg = <0x0000 0 0 0 0>;
	};
};
```

looks good?

Reply:
```
[    1.990415] pci 0000:01:00.0: [14c3:7615] type 00 class 0x000280                                                                    
[    1.990465] pci 0000:01:00.0: reg 0x10: [mem 0x00000000-0x000fffff 64bit]                                                           
[    1.990732] pci_bus 0000:01: busn_res: [bus 01-ff] end is updated to 01                                                             
[    1.990913] pci 0000:02:00.0: [14c3:7615] type 00 class 0x000280                                                                    
[    1.990963] pci 0000:02:00.0: reg 0x10: [mem 0x00000000-0x000fffff 64bit]                                                           
[    1.991236] pci_bus 0000:02: busn_res: [bus 02-ff] end is updated to 02            
```

ok, the pciid seem to be correct. But ```0000:01:00.0``` and ```0000:02:00.0``` makes me think that it could be pcie1 and pcie2 (and not pcie0).

Can you please run a little test and see if:

```
&pcie1 {
     status = "disabled";
};
&pcie2 {
     status = "disabled";
};
```

makes both pcie devices disappear? 

As for
```
                mediatek,mtd-eeprom = <&factory 0x8000>;
                ieee80211-freq-limit = <5000000 6000000>;
```

do you know for certain which pcie chip is which? If yes, then it's most likely "Okay". Although, the mt76 driver doesn't support the MT7615 yet, so the "mediatek,mtd-eeprom" might be a bit wishful thinking. Anyway, more tomorrow.

Reply:
I've spent most of the day playing around with unholy pci.

In response to your questions/suggestions:

1) `0000:01:00.0 `and `0000:02:00.0` also had me wondering whether &pcie1 and &pcie2 would be "the right" answer. I've tested and can verify that the correct answer indeed is pcie0 and pcie1. (If you change the dts to have pcie1 and pcie2 the path to one of the mt7615s (ie pcie2) ends up being unrecognized (ie, same as if you didn't have any section for it in the dts file).

2) putting `status = "disabled";` seems to make absolutely no difference (as far as I can see, including lspci). That's the result of testing many combinations.

3) As for whether I know which pci device maps to 2.4ghz and which to 5ghz, yes I finally do. I loaded the binary-only "leaked" mt7615 modules, found the pci device in /sys/devices and did an `echo 1 >> remove`. Shortly before the kernel panics, it tells you which one it was trying to disable. It's as I had in the dts files: pcie0 is 2.4ghz and pcie1 is 5ghz.

4) "mediatek,mtd-eeprom" most likely *is* wishful thinking for the time being, but since we're adding the nodes for the (unsupported) pci devices, I think it's right to configure them properly (and not have to have follow-up PRs fixing up a bunch of dts files once mt7615 is supported). It also serves as a useful comment (and the person who might be doing those follow-up PRs may not have access to every device with an mt7615 chip). BTW, I've verified that those addresses are correct.

5) As for the complete boot message: I can't for the life of me figure out why it displayed the device IDs yesterday and refuses to do so today. Here's the relevant portion of the boot message:
```

[    0.414689] pull PCIe RST: RALINK_RSTCTRL = 4000000                                                                                 
[    0.719884] release PCIe RST: RALINK_RSTCTRL = 7000000                                                                              
[    0.724931] ***** Xtal 40MHz *****                                                                                                  
[    0.728287] release PCIe RST: RALINK_RSTCTRL = 7000000                                                                              
[    0.733384] Port 0 N_FTS = 1b105000                                                                                                 
[    0.736852] Port 1 N_FTS = 1b105000                                                                                                 
[    0.740299] Port 2 N_FTS = 1b102800                                                                                                 
[    1.895433] PCIE2 no card, disable it(RST&CLK)                                                                                      
[    1.899791]  -> 21007f2                                                                                                             
[    1.902199] PCIE0 enabled                                                                                                           
[    1.904787] PCIE1 enabled                                                                                                           
[    1.907391] PCI host bridge /pcie@1e140000 ranges:                                                                                  
[    1.912164]  MEM 0x0000000060000000..0x000000006fffffff                                                                             
[    1.917320]   IO 0x000000001e160000..0x000000001e16ffff                                                                             
[    1.922525] PCI coherence region base: 0xbfbf8000, mask/settings: 0x60000000                                                        
[    1.938442] mt7621_gpio 1e000600.gpio: registering 32 gpios                                                                         
[    1.944312] mt7621_gpio 1e000600.gpio: registering 32 gpios                                                                         
[    1.950020] mt7621_gpio 1e000600.gpio: registering 32 gpios                                                                         
[    1.957576] PCI host bridge to bus 0000:00                                                                                          
[    1.961614] pci_bus 0000:00: root bus resource [mem 0x60000000-0x6fffffff]                                                          
[    1.968447] pci_bus 0000:00: root bus resource [io  0xffffffff]                                                                     
[    1.974299] pci_bus 0000:00: root bus resource [??? 0x00000000 flags 0x0]                                                           
[    1.981058] pci_bus 0000:00: No busn resource found for root bus, will use [bus 00-ff]                                              
[    1.991018] pci 0000:00:00.0: BAR 0: no space for [mem size 0x80000000]                                                             
[    1.997553] pci 0000:00:00.0: BAR 0: failed to assign [mem size 0x80000000]                                                         
[    2.004451] pci 0000:00:01.0: BAR 0: no space for [mem size 0x80000000]                                                             
[    2.011026] pci 0000:00:01.0: BAR 0: failed to assign [mem size 0x80000000]                                                         
[    2.017927] pci 0000:00:00.0: BAR 8: assigned [mem 0x60000000-0x600fffff]                                                           
[    2.024682] pci 0000:00:01.0: BAR 8: assigned [mem 0x60100000-0x601fffff]                                                           
[    2.031400] pci 0000:00:00.0: BAR 1: assigned [mem 0x60200000-0x6020ffff]                                                           
[    2.038160] pci 0000:00:01.0: BAR 1: assigned [mem 0x60210000-0x6021ffff]                                                           
[    2.044895] pci 0000:01:00.0: BAR 0: assigned [mem 0x60000000-0x600fffff 64bit]                                                     
[    2.052165] pci 0000:00:00.0: PCI bridge to [bus 01]                                                                                
[    2.057066] pci 0000:00:00.0:   bridge window [mem 0x60000000-0x600fffff]                                                           
[    2.063830] pci 0000:02:00.0: BAR 0: assigned [mem 0x60100000-0x601fffff 64bit]                                                     
[    2.071071] pci 0000:00:01.0: PCI bridge to [bus 02]                                                                                
[    2.075997] pci 0000:00:01.0:   bridge window [mem 0x60100000-0x601fffff]     
```                                                      

@chunkeey and @ynezz : as of this response I don't believe we have any remaining unresolved issues. If you don't disagree, and if there are no further comments can we move forward to accept this PR?

Reply:
Ok, then let's leave it at that.

One final question: What should be done about the 7615?
Just mention that they don't have a driver (klnd of like the 4352/4360 on bcm53xx)?

Reply:
@chunkeey I might be having late-night PCI/mtk withdrawal symptoms, but I didn't understand your last sentence ;) Are you suggesting to add a note in the commit message that mt7615 doesn't have opensource drivers (yet)?

If I understand your question to mean "what is to be done in general about mt7615" or (perhaps more relevantly) "what about xiaomi_mir3p and wifi?" the solution to the former is that mt76 is (according to some post I read somewhere) starting to think about adding support for mt7615. When that development will take place and when it will be production-ready is anyone's guess.
As for the question "will people be able to use wifi on the xiaomi_mir3p once this PR goes into openwrt or will this 'support' be theoretical?": there are binary-only kernel modules for mt7615 being periodically released (by a kind soul) here: http://nossiac.com/download/mtk-wifi-ko/ (the latest supported kernel is 4.14.93, which I've been doing my testing with). I'm unofficially releasing 4.14.93-based openwrt/mir3p builds that include the binary modules, and there are a couple of people out there who have been testing it.

So in the end, (and if I understood your question correctly) it is possible to use wifi on the mir3p and yes I hope that openwrt/mir3p will be an actively used platform.

Reply:
> @chunkeey I might be having late-night PCI/mtk withdrawal symptoms, but I didn't understand your last sentence ;) Are you suggesting to add a note in the commit message that mt7615 doesn't have opensource drivers (yet)?

Yes something along the lines with little more emphasis on the "what would/could this notice look like?". I'm aware of the forum thread about the device and I can tell you that from what I know, it's not a problem to have no driver for the device. Some bcm53xx devices are in a similar situation with unsupported wifis (BCM4352/BCM4360) and it didn't stop them from getting accepted. I guess the main question is: should the ?unknown-origin? driver be mentioned or not? I would lean to: no. Because it could cause problems for the project down the line.

Reply:
@chunkeey I agree about not mentioning the binary driver (after all, this is an *openwrt* commit, and there's no point in mentioning anything that doesn't pertain to that).

Actually, there are a number of devices already in openwrt which have mt7615: a quick grep in target/linux/ramips/dts shows at least two (with comments... most likely there are others without comments):

```
elecom_wrc-gst.dtsi:	/* WRC-xxxxGST has MT7615 for 2.4/5 GHz wifi, but it's not supported */
WRC-1167GHBK2-S.dts:	 * WRC-1167GHBK2-S has MT7615D for 2.4/5 GHz wifi,

```
I'm of the opinion that (as there are already other MT7615 devices in openwrt) the mir3p doesn't deserve special notice of any form.

Reply:
> @chunkeey I agree about not mentioning the binary driver (after all, this is an openwrt commit, and there's no point in mentioning anything that doesn't pertain to that).

oh, you can and should totally say what's not working [I think this gets lost in translation. Because in some cultures this "looks bad" and instead they prefer to call it "still being worked on" etc]. 

> I'm of the opinion that (as there are already other MT7615 devices in openwrt) the mir3p doesn't deserve special notice of any form.

Ah, I think it's important to mention it in the message for your own shake. You see, people that flash the device with OpenWrt might expect that it "works out of the box". And when even the 2.4GHz Wifi doesn't work, some get upset. Now, If you added a little note preemptively to the message you can choose whenever you point rude requests to the "commit message" or to the alternate driver. Really, this is done to protect yourself.

Anyway, if you have cooked something up, let me know.

Reply:
> oh, you can and should totally say what's not working [I think this gets lost in translation. Because in some cultures this "looks bad" and instead they prefer to call it "still being worked on" etc].

@chunkeey lol. Maybe I'd feel that way if I was personally responsible for mt76/mt7615 and/or was trying to sell people the mir3p... All I really care about is getting the base support in.. and then it's relatively easy for people to hack around the wireless if they want to.

OK, back to business. I think we're agreed that mention of "alternate driver"(s) doesn't belong in an openwrt commit message. There's plenty of traffic about that on the Forum thread. How about I add a link to the MIR3P forum page and mention that openwrt doesn't support mt7615 out of the box?

Reply:
> @chunkeey I might be having late-night PCI/mtk withdrawal symptoms, but I didn't understand your last sentence ;) Are you suggesting to add a note in the commit message that mt7615 doesn't have opensource drivers (yet)?

Yes, please add a notice about what does not work in the commit message. Also for @chunkeey this is the main reason we should want a notice.

The people adding the device specifications for supported devices in the wiki Table of Hardware https://openwrt.org/toh/views/toh_extended_all (it's usually tmomas) will need this information as this is important to know for the users.

Reply:
> OK, back to business. I think we're agreed that mention of "alternate driver"(s) doesn't belong in an openwrt commit message. There's plenty of traffic about that on the Forum thread. How about I add a link to the MIR3P forum page and mention that openwrt doesn't support mt7615 out of the box?

Yes. that would work really nicely. :+1: